### PR TITLE
[WIP] MembershipType/ID in URL

### PR DESCRIPTION
**Pull request by joshhunt**
## Changes

 * Current user 'state' is now stored in the URL. 
   * All user-specific views are now nested inside of CharacterRoutes
   * CharacterRoutes is what fetched the profile
 * When 'highlighting' a record (via url), that record will always be shown, even if it's completed and 'hide completed' has been enabled.

## So, about this PR

The main user feature goal here is to revert back to have membership ID in the URL so users can bookmark and share pages better.

The git diff might seem really big and overwhelming, but architecturally the ideal here is relatively straight forward:

 - Simplify `<App />` by including a single set of routes, regardless of signed in state
    - Routes directly in `<App />` should be profile-agnostic, like global settings and about pages
 - Split all profile/character specific routes into its own component/subrouter, `<CharacterRoutes />`
 - Centralise profile-fetching code to something higher up in the tree. Current, this is in `<CharacterRoutes />`, thought it might make sense to do that in `<App />` instead.

Most of the changes in all other components (mostly the views) were just about fixing URLs to link to the new pages containing the membershipType/ID.

A few of the other misc changes that had to be done, or I took the opportunity to clean up:

 - `<ProfileLink />` and `<ProfileNavLink />` that wrap react-router-dom's `<Link />` and `<NavLink />` shallowly to provide a convenient way to link to pages of the current character without having to explicitly know their membershipID/type when calling it.
    - This uses [React context](https://reactjs.org/docs/context.html) to make the user object accessible inside `<ProfileLink />` without having to manually pass it all the way down.
 - CharacterRoutes fetches the profile for the current url (membershipType/ID is gotten from the URL params), and sends it to App.js using its `setUserReponse` method.
    - I broke out the profile fetching code into a separate file to maintain readability and promote code reuse.
 - All the Loading view view logic has been broken out into a separate component, and was simplified.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -252,228 +247,48 @@ class App extends Component {
...
-          <div className='view' id='loading'>
-            <div className='logo-feature'>
-              <div className='device'>
-                <span className='destiny-clovis_bray_device' />
-              </div>
+      return <Loading state={this.state.manifest.state} />;
+    }
+
+    return (
+      <UserContext.Provider value={this.state.user}>
```
Comment (joshhunt) [inline src/App.js]: this UserContext is from [React Context](https://reactjs.org/docs/context.html) and is required to pass the user directly down into `<ProfileLink />`

**Comment by justrealmilk** [comment]
I slap merge so hard. Have to sort out the conflicts one by one

**Comment by justrealmilk** [comment]
_KFC Colonel voice_ Kindly replace _CharacterSelectRedirect_ with a plain ol' inline _Redirect_ as previously done did

**Comment by joshhunt** [comment]
Okay, this is... looking good? Maybe?

**Pull request merged**